File: app/src/main/java/com/example/fondodepantalla/FragmentosAdministrador/RegistrarAdmin.java
package com.example.fondodepantalla.FragmentosAdministrador;

import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.fragment.app.Fragment;

import com.anton46.stepsview.StepsView;
import com.cloudinary.android.MediaManager;
import com.cloudinary.android.callback.ErrorInfo;
import com.cloudinary.android.callback.UploadCallback;
import com.example.fondodepantalla.R;
import com.github.gcacace.signaturepad.views.SignaturePad;
import com.google.firebase.firestore.FieldValue;
import com.google.firebase.firestore.FirebaseFirestore;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.util.Map;

public class RegistrarAdmin extends Fragment {
    private ProgressBar progressBarConfirmacion;

    private StepsView stepsView;
    private FirebaseFirestore db;
    private String taskId;
    private Button btnNextStep, btnSubirEvidencia, btnEnviarConfirmacion;
    private int currentStep = 0;

    private CheckBox checkExito;
    private EditText etComentario;
    private SignaturePad signaturePad;

    private static final int PICK_IMAGES_REQUEST = 101;

    private final String[] labels = {"Inicio", "st1", "st2", "st3", "Completa"};

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_registro_admin, container, false);

        stepsView = view.findViewById(R.id.stepsView);
        btnNextStep = view.findViewById(R.id.btnNextStep);
        btnSubirEvidencia = view.findViewById(R.id.btnSubirEvidencia);
        db = FirebaseFirestore.getInstance();

        // Formularios de confirmación
        btnEnviarConfirmacion = view.findViewById(R.id.btnEnviarConfirmacion);
        checkExito = view.findViewById(R.id.checkExito);
        etComentario = view.findViewById(R.id.etComentario);
        signaturePad = view.findViewById(R.id.signaturePad);

        if (getArguments() != null) {
            taskId = getArguments().getString("taskId");
        }

        initStepsView();

        if (taskId != null) loadTaskProgress();
        else Toast.makeText(getContext(), "ID de tarea nulo", Toast.LENGTH_SHORT).show();

        setupButton();

        // Subir evidencia
        btnSubirEvidencia.setOnClickListener(v -> {
            Intent intent = new Intent();
            intent.setType("image/*");
            intent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true);
            intent.setAction(Intent.ACTION_GET_CONTENT);
            startActivityForResult(Intent.createChooser(intent, "Selecciona evidencia"), PICK_IMAGES_REQUEST);
        });

        // Enviar confirmación
        btnEnviarConfirmacion.setOnClickListener(v -> enviarConfirmacion());

        progressBarConfirmacion = view.findViewById(R.id.progressBarConfirmacion);

        return view;
    }

    private void initStepsView() {
        stepsView.setLabels(labels)
                .setBarColorIndicator(getResources().getColor(R.color.pastel_gray))
                .setProgressColorIndicator(getResources().getColor(R.color.pastel_purple))
                .setLabelColorIndicator(getResources().getColor(R.color.black))
                .setCompletedPosition(currentStep)
                .drawView();
    }

    private void loadTaskProgress() {
        db.collection("tareas").document(taskId)
                .get()
                .addOnSuccessListener(doc -> {
                    if (doc.exists()) {
                        Long progreso = doc.getLong("progress");
                        if (progreso != null) {
                            currentStep = progreso.intValue();
                            refreshStepsView();
                            toggleEvidenciaButton();
                        }
                    }
                });
    }

    private void updateProgress(int progress) {
        if (taskId == null) return;
        db.collection("tareas").document(taskId)
                .update("progress", progress)
                .addOnSuccessListener(aVoid -> {
                    currentStep = progress;
                    refreshStepsView();
                    toggleEvidenciaButton();
                    Toast.makeText(getContext(), "Avanzaste a: " + labels[progress], Toast.LENGTH_SHORT).show();
                })
                .addOnFailureListener(e -> Toast.makeText(getContext(), "Error al actualizar progreso", Toast.LENGTH_SHORT).show());
    }

    private void refreshStepsView() {
        stepsView.setCompletedPosition(currentStep).drawView();
    }

    private void setupButton() {
        btnNextStep.setOnClickListener(v -> {
            if (currentStep < labels.length - 1) updateProgress(currentStep + 1);
            else Toast.makeText(getContext(), "Ya estás en el último paso", Toast.LENGTH_SHORT).show();
        });
    }

    private void toggleEvidenciaButton() {
        if (btnSubirEvidencia != null)
            btnSubirEvidencia.setVisibility(currentStep == 4 ? View.VISIBLE : View.GONE);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == PICK_IMAGES_REQUEST && resultCode == getActivity().RESULT_OK) {
            if (data.getClipData() != null) { // varias imágenes
                int count = data.getClipData().getItemCount();
                for (int i = 0; i < count; i++) {
                    Uri imageUri = data.getClipData().getItemAt(i).getUri();
                    subirArchivoCloudinary(imageUri);
                }
            } else if (data.getData() != null) { // una sola imagen
                Uri imageUri = data.getData();
                subirArchivoCloudinary(imageUri);
            }
        }
    }

    private void subirArchivoCloudinary(Uri fileUri) {
        if (taskId == null) return;

        MediaManager.get().upload(fileUri)
                .option("folder", "evidencias/" + taskId)
                .option("use_filename", true)
                .option("unique_filename", true)
                .option("resource_type", "image")
                .callback(new UploadCallback() {
                    @Override
                    public void onStart(String requestId) {}

                    @Override
                    public void onProgress(String requestId, long bytes, long totalBytes) {}

                    @Override
                    public void onSuccess(String requestId, Map resultData) {
                        String url = (String) resultData.get("secure_url");
                        db.collection("tareas").document(taskId)
                                .update("evidencias", FieldValue.arrayUnion(url));
                        Toast.makeText(getContext(), "Imagen subida correctamente", Toast.LENGTH_SHORT).show();

                        // Mostrar formulario de confirmación
                        View layoutConfirmacion = getView().findViewById(R.id.layoutConfirmacion);
                        layoutConfirmacion.setVisibility(View.VISIBLE);
                    }

                    @Override
                    public void onError(String requestId, ErrorInfo error) {
                        Toast.makeText(getContext(), "Error al subir imagen: " + error.getDescription(), Toast.LENGTH_SHORT).show();
                    }

                    @Override
                    public void onReschedule(String requestId, ErrorInfo error) {}
                })
                .dispatch();
    }

    private void enviarConfirmacion() {
        boolean exito = checkExito.isChecked();
        String comentario = etComentario.getText().toString();
        Bitmap firma = signaturePad.getSignatureBitmap();

        if (!exito) {
            Toast.makeText(getContext(), "Marca que se completó la tarea para continuar", Toast.LENGTH_SHORT).show();
            return;
        }

        // Mostrar ProgressBar
        progressBarConfirmacion.setVisibility(View.VISIBLE);
        btnEnviarConfirmacion.setEnabled(false); // Evitar múltiples clicks

        Uri firmaUri = bitmapToUri(firma);
        if (firmaUri == null) {
            Toast.makeText(getContext(), "Error al generar archivo de firma", Toast.LENGTH_SHORT).show();
            progressBarConfirmacion.setVisibility(View.GONE);
            btnEnviarConfirmacion.setEnabled(true);
            return;
        }

        MediaManager.get().upload(firmaUri)
                .option("folder", "firmas/" + taskId)
                .option("use_filename", true)
                .option("unique_filename", true)
                .callback(new UploadCallback() {
                    @Override
                    public void onStart(String requestId) {}

                    @Override
                    public void onProgress(String requestId, long bytes, long totalBytes) {}

                    @Override
                    public void onSuccess(String requestId, Map resultData) {
                        String firmaUrl = (String) resultData.get("secure_url");

                        db.collection("tareas").document(taskId)
                                .update(
                                        "confirmacionExito", exito,
                                        "comentarioCliente", comentario,
                                        "firmaCliente", firmaUrl,
                                        "fechaConfirmacion", FieldValue.serverTimestamp()
                                )
                                .addOnSuccessListener(aVoid -> {
                                    Toast.makeText(getContext(), "Confirmación enviada", Toast.LENGTH_SHORT).show();

                                    // Limpiar formulario
                                    checkExito.setChecked(false);
                                    etComentario.setText("");
                                    signaturePad.clear();

                                    // Ocultar formulario y ProgressBar
                                    View layoutConfirmacion = getView().findViewById(R.id.layoutConfirmacion);
                                    layoutConfirmacion.setVisibility(View.GONE);
                                    progressBarConfirmacion.setVisibility(View.GONE);
                                    btnEnviarConfirmacion.setEnabled(true);
                                })
                                .addOnFailureListener(e -> {
                                    Toast.makeText(getContext(), "Error al guardar confirmación", Toast.LENGTH_SHORT).show();
                                    progressBarConfirmacion.setVisibility(View.GONE);
                                    btnEnviarConfirmacion.setEnabled(true);
                                });
                    }

                    @Override
                    public void onError(String requestId, ErrorInfo error) {
                        Toast.makeText(getContext(), "Error al subir firma: " + error.getDescription(), Toast.LENGTH_SHORT).show();
                        progressBarConfirmacion.setVisibility(View.GONE);
                        btnEnviarConfirmacion.setEnabled(true);
                    }

                    @Override public void onReschedule(String requestId, ErrorInfo error) {}
                })
                .dispatch();
    }

    // Convierte Bitmap a Uri usando archivo temporal (recomendado)
    private Uri bitmapToUri(Bitmap bitmap) {
        try {
            File file = new File(getContext().getCacheDir(), "firma_" + System.currentTimeMillis() + ".png");
            FileOutputStream out = new FileOutputStream(file);
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, out);
            out.flush();
            out.close();
            return Uri.fromFile(file);
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

}

File: app/src/main/java/com/example/fondodepantalla/FragmentosAdministrador/RegistrarAdmin.kt
package com.example.fondodepantalla.FragmentosAdministrador

import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.view.ViewGroup
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.animateContentSize
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.viewinterop.AndroidView
import androidx.fragment.app.Fragment
import com.cloudinary.android.MediaManager
import com.cloudinary.android.callback.ErrorInfo
import com.cloudinary.android.callback.UploadCallback
import com.github.gcacace.signaturepad.views.SignaturePad
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import java.io.File
import java.io.FileOutputStream

class RegistrarAdmin : Fragment() {

    private var taskId: String? = null
    private val db = FirebaseFirestore.getInstance()
    private var evidenciaSubida by mutableStateOf(false)

    private val pickImageLauncher =
        registerForActivityResult(ActivityResultContracts.GetContent()) { uri: Uri? ->
            uri?.let {
                // Subir evidencia a Cloudinary
                MediaManager.get().upload(uri)
                    .option("folder", "evidencias/${taskId ?: "tarea"}")
                    .option("use_filename", true)
                    .option("unique_filename", true)
                    .callback(object : UploadCallback {
                        override fun onStart(requestId: String?) {}
                        override fun onProgress(requestId: String?, bytes: Long, totalBytes: Long) {}
                        override fun onSuccess(requestId: String?, resultData: Map<*, *>?) {
                            Toast.makeText(requireContext(), "Evidencia subida", Toast.LENGTH_SHORT).show()
                            evidenciaSubida = true
                        }

                        override fun onError(requestId: String?, error: ErrorInfo?) {
                            Toast.makeText(
                                requireContext(),
                                "Error al subir evidencia: ${error?.description}",
                                Toast.LENGTH_SHORT
                            ).show()
                        }

                        override fun onReschedule(requestId: String?, error: ErrorInfo?) {}
                    }).dispatch()
            }
        }

    override fun onCreateView(
        inflater: android.view.LayoutInflater,
        container: android.view.ViewGroup?,
        savedInstanceState: Bundle?
    ): android.view.View {
        taskId = arguments?.getString("taskId")

        return ComposeView(requireContext()).apply {
            setContent {
                RegistrarAdminScreen()
            }
        }
    }

    @Composable
    private fun RegistrarAdminScreen() {
        val labels = listOf("Inicio", "st1", "st2", "st3", "Completa")
        var currentStep by remember { mutableStateOf(0) }
        var checkExito by remember { mutableStateOf(false) }
        var comentario by remember { mutableStateOf("") }
        var isUploading by remember { mutableStateOf(false) }
        val scrollState = rememberScrollState()
        val context = LocalContext.current
        val signaturePadRef = remember { mutableStateOf<SignaturePad?>(null) }

        // Cargar progreso inicial
        LaunchedEffect(taskId) {
            taskId?.let {
                db.collection("tareas").document(it)
                    .get()
                    .addOnSuccessListener { doc ->
                        val progreso = doc.getLong("progress")?.toInt() ?: 0
                        currentStep = progreso
                    }
            }
        }

        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(scrollState)
                .padding(16.dp),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            // Steps
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                labels.forEachIndexed { index, label ->
                    Column(horizontalAlignment = Alignment.CenterHorizontally) {
                        val color by animateColorAsState(
                            targetValue = if (index <= currentStep) Color(0xFF9C27B0) else Color.LightGray
                        )
                        Box(
                            modifier = Modifier
                                .size(40.dp)
                                .background(color = color, shape = MaterialTheme.shapes.small)
                        ) {
                            Text(
                                text = when (index) {
                                    0 -> "🏁"
                                    1 -> "📌"
                                    2 -> "🚚"
                                    3 -> "🔧"
                                    else -> "✅"
                                },
                                modifier = Modifier.align(Alignment.Center),
                                fontSize = 20.sp
                            )
                        }
                        Spacer(modifier = Modifier.height(4.dp))
                        Text(text = label, fontSize = 12.sp)
                    }
                }
            }

            Spacer(modifier = Modifier.height(24.dp))

            val buttonColor by animateColorAsState(
                targetValue = if (currentStep < labels.lastIndex) Color(0xFF64B5F6) else Color.Gray
            )

            Button(
                onClick = {
                    if (currentStep < labels.lastIndex) {
                        currentStep++
                        taskId?.let { id ->
                            db.collection("tareas").document(id)
                                .update("progress", currentStep)
                                .addOnSuccessListener {
                                    Toast.makeText(
                                        context,
                                        "Avanzaste a: ${labels[currentStep]}",
                                        Toast.LENGTH_SHORT
                                    ).show()
                                }
                        }
                    } else {
                        Toast.makeText(context, "Ya estás en el último paso", Toast.LENGTH_SHORT).show()
                    }
                },
                colors = ButtonDefaults.buttonColors(containerColor = buttonColor),
                modifier = Modifier.fillMaxWidth(0.6f).height(50.dp)
            ) {
                Text("Confirmar paso", color = Color.White, fontSize = 16.sp)
            }

            Spacer(modifier = Modifier.height(16.dp))

            // Botón subir evidencia
            if (currentStep == labels.lastIndex && !evidenciaSubida) {
                Button(
                    onClick = { pickImageLauncher.launch("image/*") },
                    colors = ButtonDefaults.buttonColors(containerColor = Color(0xFFFFAAA5)),
                    modifier = Modifier.fillMaxWidth(0.6f)
                ) {
                    Text("Subir evidencia", color = Color.Black, fontSize = 16.sp)
                }
            }

            Spacer(modifier = Modifier.height(24.dp))

            // Formulario de confirmación
            AnimatedVisibility(visible = evidenciaSubida) {
                Card(
                    modifier = Modifier.fillMaxWidth().animateContentSize(),
                    shape = MaterialTheme.shapes.medium,
                    elevation = CardDefaults.cardElevation(defaultElevation = 8.dp)
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Text(
                            "¿Se realizó el trabajo solicitado con éxito?",
                            fontWeight = androidx.compose.ui.text.font.FontWeight.Bold,
                            fontSize = 16.sp
                        )
                        Spacer(modifier = Modifier.height(8.dp))

                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Checkbox(
                                checked = checkExito,
                                onCheckedChange = { checkExito = it }
                            )
                            Spacer(modifier = Modifier.width(8.dp))
                            Text("Sí, se completó")
                        }

                        Spacer(modifier = Modifier.height(12.dp))

                        OutlinedTextField(
                            value = comentario,
                            onValueChange = { comentario = it },
                            label = { Text("Comentario (opcional)") },
                            modifier = Modifier.fillMaxWidth(),
                            minLines = 3,
                            maxLines = 5
                        )

                        Spacer(modifier = Modifier.height(12.dp))

                        AndroidView(factory = { ctx ->
                            SignaturePad(ctx, null).apply {
                                setSaveEnabled(false)
                                signaturePadRef.value = this
                                setBackgroundColor(android.graphics.Color.LTGRAY)
                            }
                        }, modifier = Modifier.fillMaxWidth().height(200.dp))

                        Spacer(modifier = Modifier.height(12.dp))

                        Row(
                            horizontalArrangement = Arrangement.spacedBy(8.dp),
                            modifier = Modifier.fillMaxWidth()
                        ) {
                            Button(
                                onClick = { signaturePadRef.value?.clear() },
                                colors = ButtonDefaults.buttonColors(containerColor = Color.Gray),
                                modifier = Modifier.weight(1f)
                            ) {
                                Text("Borrar firma", color = Color.White)
                            }

                            Button(
                                onClick = {
                                    val firmaBitmap = signaturePadRef.value?.signatureBitmap
                                    if (firmaBitmap != null && checkExito) {
                                        isUploading = true
                                        taskId?.let { id ->
                                            val uri = bitmapToUri(firmaBitmap)
                                            if (uri != null) {
                                                MediaManager.get().upload(uri)
                                                    .option("folder", "firmas/$id")
                                                    .option("use_filename", true)
                                                    .option("unique_filename", true)
                                                    .callback(object : UploadCallback {
                                                        override fun onStart(requestId: String?) {}
                                                        override fun onProgress(requestId: String?, bytes: Long, totalBytes: Long) {}
                                                        override fun onSuccess(requestId: String?, resultData: Map<*, *>?) {
                                                            val firmaUrl = resultData?.get("secure_url") as? String
                                                            db.collection("tareas").document(id)
                                                                .update(
                                                                    mapOf(
                                                                        "confirmacionExito" to checkExito,
                                                                        "comentarioCliente" to comentario,
                                                                        "firmaCliente" to firmaUrl,
                                                                        "fechaConfirmacion" to FieldValue.serverTimestamp()
                                                                    )
                                                                ).addOnSuccessListener {
                                                                    Toast.makeText(context, "¡Tarea completada!", Toast.LENGTH_SHORT).show()
                                                                    // Reset
                                                                    signaturePadRef.value?.clear()
                                                                    checkExito = false
                                                                    comentario = ""
                                                                    evidenciaSubida = false
                                                                    isUploading = false
                                                                    parentFragmentManager.beginTransaction().apply {
                                                                        // Remueve el fragmento actual
                                                                        remove(this@RegistrarAdmin)

                                                                        // Agrega ListaAdmin al mismo contenedor del fragmento actual
                                                                        val parent = view?.parent as? ViewGroup
                                                                        parent?.let { add(it.id, ListaAdmin()) }
                                                                        commit()
                                                                    }
                                                                }
                                                        }

                                                        override fun onError(requestId: String?, error: ErrorInfo?) {
                                                            Toast.makeText(context, "Error al subir firma: ${error?.description}", Toast.LENGTH_SHORT).show()
                                                            isUploading = false
                                                        }

                                                        override fun onReschedule(requestId: String?, error: ErrorInfo?) {}
                                                    }).dispatch()
                                            }
                                        }
                                    } else {
                                        Toast.makeText(context, "Marca que se completó la tarea y firma antes de enviar", Toast.LENGTH_SHORT).show()
                                    }
                                },
                                colors = ButtonDefaults.buttonColors(containerColor = Color(0xFF64B5F6)),
                                modifier = Modifier.weight(1f)
                            ) {
                                Text("Enviar", color = Color.White)
                            }
                        }

                        if (isUploading) {
                            Spacer(modifier = Modifier.height(12.dp))
                            CircularProgressIndicator(modifier = Modifier.align(Alignment.CenterHorizontally))
                        }
                    }
                }
            }
        }
    }

    private fun bitmapToUri(bitmap: Bitmap): Uri? {
        return try {
            val file = File(requireContext().cacheDir, "firma_${System.currentTimeMillis()}.png")
            FileOutputStream(file).use { out ->
                bitmap.compress(Bitmap.CompressFormat.PNG, 100, out)
            }
            Uri.fromFile(file)
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }
}
